### PR TITLE
refactor: validate should allow for execution context

### PR DESCRIPTION
**Pull request by natedanner**
Problem: 
Validation doesn't use ExecutionContext on recipes. There is a need to be able to add customizations through the ExecutionContext during recipe validation. 

Solution: 
Enhance base recipe execution to allow for execution context. 

Outcome: 
Ability to use ExecutionContext for recipe validation.

### CODE REVIEW

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java [hunk 1]
```diff
@@ -73,10 +74,13 @@ public String getDescription() {
     Validated gradleWrapper;
 
     @Override
-    public Validated validate() {
+    public Validated validate(ExecutionContext ctx) {
         if (gradleWrapper == null) {
-            gradleWrapper = super.validate().and(GradleWrapper.validate(version, distribution,
-                    new HttpUrlConnectionSender(Duration.ofSeconds(3), Duration.ofSeconds(10))));
+            HttpSender httpSender = ctx.getMessage("httpSender");
```
Comment (jkschneider) [inline rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java:79]: Can we namespace this property `org.openrewrite.httpSender`?

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/UpdateGradleWrapper.java [hunk 2]
```diff
@@ -71,10 +72,13 @@ public String getDescription() {
     Validated gradleWrapper;
 
     @Override
-    public Validated validate() {
+    public Validated validate(ExecutionContext ctx) {
         if (gradleWrapper == null) {
-            gradleWrapper = super.validate().and(GradleWrapper.validate(version, distribution,
-                    new HttpUrlConnectionSender(Duration.ofSeconds(3), Duration.ofSeconds(10))));
+            HttpSender httpSender = ctx.getMessage("httpSender");
```
Comment (jkschneider) [inline rewrite-gradle/src/main/java/org/openrewrite/gradle/UpdateGradleWrapper.java:77]: Similar namespacing.

**Review by sambsnyd — CHANGES_REQUESTED** [review]

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java (hunk 1 above)
Comment (natedanner) [reply rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java]: Thanks. View added.

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/UpdateGradleWrapper.java (hunk 2 above)
Comment (natedanner) [reply rewrite-gradle/src/main/java/org/openrewrite/gradle/UpdateGradleWrapper.java]: fixed in view

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java [hunk 3]
```diff
@@ -73,10 +74,13 @@ public String getDescription() {
...
+            if (httpSender == null) {
+                httpSender = new HttpUrlConnectionSender(Duration.ofSeconds(3), Duration.ofSeconds(10));
+            }
```
[the remark this answers is not in the document]
Comment (natedanner) [reply rewrite-gradle/src/main/java/org/openrewrite/gradle/AddGradleWrapper.java]: @sambsnyd  Added view. Thank you!

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/GradleExecutionContextView.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.gradle;
+
+import org.openrewrite.DelegatingExecutionContext;
+import org.openrewrite.ExecutionContext;
+import org.openrewrite.ipc.http.HttpUrlConnectionSender;
+
+public class GradleExecutionContextView extends DelegatingExecutionContext {
```
Comment (sambsnyd) [inline rewrite-gradle/src/main/java/org/openrewrite/gradle/GradleExecutionContextView.java:22]:
```suggestion
public class HttpSenderExecutionContextView extends DelegatingExecutionContext {
```

This view isn't really Gradle-specific. All Recipes which need to issue HTTP requests, including Maven recipes, will be updated to get their `HttpSender` from execution context

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/GradleExecutionContextView.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+    private static final String HTTP_SENDER = "httpSender";
```
Comment (sambsnyd) [inline rewrite-gradle/src/main/java/org/openrewrite/gradle/GradleExecutionContextView.java:24]:
```suggestion
    private static final String HTTP_SENDER = "org.openrewrite.httpSender";
```

**Review by sambsnyd — CHANGES_REQUESTED** [review]

File: rewrite-gradle/src/main/java/org/openrewrite/gradle/GradleExecutionContextView.java (hunk 4 above)
Comment (natedanner) [reply rewrite-gradle/src/main/java/org/openrewrite/gradle/GradleExecutionContextView.java]: Renamed and moved this into rewrite-core.

**Review by sambsnyd — APPROVED** [review]

**Pull request merged**